editor/src/clj/editor/process.clj

```diff
@@ -13,31 +13,157 @@
 ;; specific language governing permissions and limitations under the License.
 
 (ns editor.process
-  (:require [clojure.java.io :as io])
-  (:import [java.lang Process ProcessBuilder]))
+  (:require [clojure.java.io :as io]
+            [clojure.string :as string])
+  (:import [java.lang Process ProcessBuilder ProcessBuilder$Redirect]
+           [java.util List]
+           [java.io InputStream StringWriter OutputStream]))
 
 (set! *warn-on-reflection* true)
 
-(defn start! ^Process [^String command args {:keys [directory env removeenv redirect-error-stream?]
-                                             :or {redirect-error-stream? false}
-                                             :as opts}]
-  (let [pb (ProcessBuilder. ^"[Ljava.lang.String;" (into-array String (list* command args)))]
-    (when directory
-      (.directory pb directory))
+(defn- ->redirect
+  ^ProcessBuilder$Redirect [x]
+  (case x
+    :pipe ProcessBuilder$Redirect/PIPE
+    :inherit ProcessBuilder$Redirect/INHERIT
+    :discard ProcessBuilder$Redirect/DISCARD
+    x))
+
+(defn ^{:arglists '([& command+args] [opts & command+args])} start
```

Comment:
Don't we want most of the functions in this file to have the`!` suffix? I'd think calling these from a `swap!` function would be bad?